lib/Sema/TypeCheckMacros.cpp

```diff
@@ -25,11 +25,71 @@
 
 using namespace swift;
 
+extern "C" void *swift_ASTGen_lookupMacro(const char *macroName);
+
+extern "C" void swift_ASTGen_destroyMacro(void *macro);
+
+extern "C" void
+swift_ASTGen_getMacroEvaluationContext(const void *sourceFile,
+                                       void *declContext, void *astContext,
+                                       void *macro, void **evaluationContext);
+
 extern "C" ptrdiff_t swift_ASTGen_evaluateMacro(
     void *sourceFile, const void *sourceLocation,
     const char **evaluatedSource, ptrdiff_t *evaluatedSourceLength);
 
+extern "C" void
+swift_ASTGen_getMacroTypeSignature(void *macro, const char **genericSignature,
+                                   ptrdiff_t *genericSignatureLength);
+
 #if SWIFT_SWIFT_PARSER
+
+StructDecl *
+swift::macro_context::lookup(StringRef macroName, DeclContext *DC) {
+  auto &ctx = DC->getASTContext();
+
+  return ctx.getOrCreateASTGenMacroContext(
+      macroName, [&](StringRef macroName) -> StructDecl * {
+        auto *macro = swift_ASTGen_lookupMacro(macroName.str().c_str());
+        if (!macro)
+          return nullptr;
+
+        const char *evaluatedSource;
+        ptrdiff_t evaluatedSourceLength;
+        swift_ASTGen_getMacroTypeSignature(macro, &evaluatedSource,
+                                           &evaluatedSourceLength);
+
+        // Create a new source buffer with the contents of the macro's
+        // signature.
+        SourceManager &sourceMgr = ctx.SourceMgr;
+        std::string bufferName;
+        {
+          llvm::raw_string_ostream out(bufferName);
+          out << "Macro signature of #" << macroName;
+        }
+        auto macroBuffer = llvm::MemoryBuffer::getMemBuffer(
+            StringRef(evaluatedSource, evaluatedSourceLength), bufferName);
+        unsigned macroBufferID =
+            sourceMgr.addNewSourceBuffer(std::move(macroBuffer));
+        auto macroSourceFile = new (ctx) SourceFile(
+            *DC->getParentModule(), SourceFileKind::Library, macroBufferID);
+        // Make sure implicit imports are resolved in this file.
+        performImportResolution(*macroSourceFile);
+
+        auto *start = sourceMgr.getLocForBufferStart(macroBufferID)
+                          .getOpaquePointerValue();
+        void *context = nullptr;
+        swift_ASTGen_getMacroEvaluationContext(
+            (const void *)start, (void *)(DeclContext *)macroSourceFile,
+            (void *)&ctx, macro, &context);
+
+        ctx.addCleanup([macro]() { swift_ASTGen_destroyMacro(macro); });
+        return dyn_cast<StructDecl>((Decl *)context);
+      });
+}
+
+#endif // SWIFT_SWIFT_PARSER
```

Comment:
I think this has to stay at the end of the file, so none of this code is compiled in when `SWIFT_SWIFT_PARSER` is not defined to `1`.